File: s7/ProgProj/testing/golang/bakkersonline.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"time"

	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/chromedp"
)

func main() {
	// Create a new context
	ctx, cancel := chromedp.NewContext(context.Background())
	defer cancel()
	fmt.Print("running")

	// Navigate to the website
	err := chromedp.Run(ctx, chromedp.Navigate("https://webshop.broodenbanketmartens.be/be-nl/martens"))
	if err != nil {
		panic(err)
	}

	// Wait for category links to be present
	var categoryLinks []*cdp.Node
	err = chromedp.Run(ctx, chromedp.Nodes(".list li a.category-link", &categoryLinks, chromedp.ByQueryAll))
	if err != nil {
		panic(err)
	}

	// Extract category URLs
	var categoryURLs []string
	for _, link := range categoryLinks {
		categoryURLs = append(categoryURLs, link.AttributeValue("href"))
	}

	// List to store scraped data
	var allProducts []map[string]interface{}

	// Loop through each category URL and scrape products
	for _, categoryURL := range categoryURLs {
		fmt.Println(categoryURL)
		// Navigate to the category URL
		categoryURL = "https://webshop.broodenbanketmartens.be" + categoryURL
		err := chromedp.Run(ctx, chromedp.Navigate(categoryURL))
		if err != nil {
			fmt.Println("Failed to navigate to category URL:", err)
			continue
		}

		// Wait for product cards to be visible
		var productCards []*cdp.Node
		err = chromedp.Run(ctx,
			chromedp.Nodes(".blocks.product-card", &productCards, chromedp.ByQueryAll),
			chromedp.WaitVisible(".blocks.product-card"),
			chromedp.NodeReady(".blocks.product-card"),
		)

		// Extract and store product names, images, and prices for this category
		var categoryProducts []map[string]interface{}
		for _, card := range productCards {
			fmt.Println(card)
			var product map[string]interface{}

			err := chromedp.Run(ctx,
				chromedp.Sleep(10*time.Second),
				chromedp.WaitVisible(fmt.Sprintf("%s .product-title", card.FullXPath())),
				chromedp.WaitVisible(fmt.Sprintf("%s .cover", card.FullXPath())),
				chromedp.WaitVisible(fmt.Sprintf("%s .price.total", card.FullXPath())),
			)
			fmt.Println("going")
			if err != nil {
				fmt.Println("Failed to wait for product card elements:", err)
				continue
			}

			// Extract product name
			var productName string
			err = chromedp.Run(ctx, chromedp.TextContent(fmt.Sprintf(`%s .product-title strong`, card.FullXPath()), &productName))
			if err != nil {
				fmt.Println("Failed to get product name:", err)
				continue
			}
			fmt.Println(productName)

			// Extract product image URL
			var productImageURL string
			err = chromedp.Run(ctx, chromedp.Evaluate(`window.getComputedStyle(document.querySelector('.cover')).backgroundImage`, &productImageURL))
			if err != nil {
				fmt.Println("Failed to get product image URL:", err)
				productImageURL = ""
			}

			// Extract product price
			var productPrice string
			err = chromedp.Run(ctx, chromedp.TextContent(fmt.Sprintf(`%s .price.total`, card.FullXPath()), &productPrice))
			if err != nil {
				fmt.Println("Failed to get product price:", err)
				continue
			}

			product = map[string]interface{}{
				"Product Name":      productName,
				"Product Image URL": productImageURL,
				"Product Price":     productPrice,
			}

			// Append the product to the category products list
			categoryProducts = append(categoryProducts, product)
		}

		// Add category products to the main list
		allProducts = append(allProducts, map[string]interface{}{categoryURL: categoryProducts})
	}

	// Convert the scraped data to JSON format
	outputJSON, err := json.MarshalIndent(allProducts, "", "    ")
	if err != nil {
		panic(err)
	}

	// Write the JSON data to an output file
	err = ioutil.WriteFile("bakkerOutput.json", outputJSON, 0644)
	if err != nil {
		panic(err)
	}

	fmt.Println("Scraped data has been successfully written to 'bakkerOutput.json'.")
}

File: s7/ProgProj/testing/selenium/bakkersonline.py
import json
from selenium import webdriver
from selenium.webdriver.common.by import By

# Initialize the WebDriver
driver = webdriver.Chrome()

# URL of the website containing the category URLs
url = "https://webshop.broodenbanketmartens.be/be-nl/martens"

# Open the website
driver.get(url)

# Wait for the category links to load (you might need to adjust the waiting time)
driver.implicitly_wait(2)

# Find all category links on the page
category_links = driver.find_elements(By.CSS_SELECTOR, 'ul.list li a.category-link')

# Extract category URLs and store them in a list
category_urls = [link.get_attribute('href') for link in category_links]

# List to store scraped data
all_products = []

# Loop through each category URL and scrape products
for category_url in category_urls:
    # Construct full URL
    full_url = f"{category_url}"
    print("Scraping URL:", full_url)
    
    # Open the website
    driver.get(full_url)
    
    # Wait for the products to load (you might need to adjust the waiting time)
    driver.implicitly_wait(2)
    
    # Find all product cards on the page
    product_cards = driver.find_elements(By.CSS_SELECTOR, 'form.blocks.product-card')

    # Extract and store product names, images, and prices for this category
    category_products = []
    for product_card in product_cards:
        product_name_element = product_card.find_element(By.CSS_SELECTOR, 'h3.product-title strong')
        product_name = product_name_element.text
        
        # Extract product image URL or assign None if it's missing
        try:
            product_image_element = product_card.find_element(By.CSS_SELECTOR, 'div.cover')
            product_image_url = product_image_element.value_of_css_property("background-image").replace('url("', '').replace('")', '')
        except Exception as e:
            print("Error occurred while extracting image:", str(e))
            product_image_url = None
        
        product_price_element = product_card.find_element(By.CSS_SELECTOR, 'strong.price.total')
        product_price = product_price_element.text
        
        # Store data in a dictionary
        product_data = {
            "Product Name": product_name,
            "Product Image URL": product_image_url,
            "Product Price": product_price
        }
        
        # Append the dictionary to the list for this category
        category_products.append(product_data)

    # Add category products to the main list
    all_products.append({category_url: category_products})

# Close the driver
driver.quit()

# Convert the list of dictionaries to JSON format
output_json = json.dumps(all_products, indent=4)

# Write the JSON data to an output file
with open("bakkerOutput.json", "w") as outfile:
    outfile.write(output_json)

print("Scraped data has been successfully written to 'bakkerOutput.json'.")
